addons/pos_self_order/controllers/kiosk_entry.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+import werkzeug
+
+from odoo import http
+from odoo.http import request
+from odoo.addons.pos_self_order.controllers.utils import reduce_privilege
+from odoo.tools.misc import get_lang
+
+
+class PosSelfKiosk(http.Controller):
+    @http.route(["/kiosk/<config_id>", "/kiosk/<config_id>/<path:subpath>"], auth="public", website=True, sitemap=True)
+    def pos_self_order_kiosk_start(self, config_id=None, access_token=None):
+        if not config_id or not access_token:
+            raise werkzeug.exceptions.NotFound()
+
+        lang = request.httprequest.cookies.get('frontend_lang')
+        if lang is not None:
+            language = get_lang(request.env, lang)
+            if language:
+                lang_code = request.env['res.lang']._lang_get_code(language.code)
+                request.lang = language
+                request.update_context(lang=lang_code)
+
+        pos_config_sudo = request.env["pos.config"].sudo().search([
+            ("id", "=", config_id),
+            ('access_token', '=', access_token)], limit=1)
+
+        if not access_token or not config_id or not pos_config_sudo:
+            raise werkzeug.exceptions.NotFound()
+
+        company = pos_config_sudo.company_id
+        user = pos_config_sudo.current_session_id.user_id
+        pos_config = reduce_privilege(pos_config_sudo, company, user)
```

Comment:
Before I did all this, there was a big problem with multi-company tax calculations, with some products using the wrong company's taxes when calculating. 

For the time being, we're going to merge this so that I can create a spot and talk to JCB so that he can review your comment.

Reply:
> For self-order and kiosk (and maybe all cases), isn't the user assigned when the session is opened? Can't we make `user_id` required, or at least check it in `reduce_privilege()`?

We'd like to make the user_id required, however, we have a mode where there is no open session for the pos.config, and in that case, self_order in mobile is only in browse mode -- where visitor can't make an order but can check the menu. In that case, `current_session_id.user_id` won't be present, which falls back to use of sudo.

Reply:
We created a task (id is 3484201) to further investigate on the use of `reduce_privilege`, `with_company` and `sudo` in the pos_self_order. @odony We might consult you soon.